### PR TITLE
dns: Add support for IPv6 and variable length IPv4 headers.

### PR DESCRIPTION
**Pull request by mauriciovasquezbernal**
Support IPv6 nameservers and extend parsing logic to support IPv4 options.

### CODE REVIEW

File: pkg/gadgets/trace/dns/tracer/bpf/dns.c [hunk 1]
```diff
@@ -109,23 +107,85 @@ SEC("socket1")
...
 
-	// Skip non-IP packets
-	if (load_half(skb, offsetof(struct ethhdr, h_proto)) != ETH_P_IP)
-		return 0;
+	// Do a first pass only to extract the port and drop the packet if it's not DNS
+	h_proto = load_half(skb, offsetof(struct ethhdr, h_proto));
+	switch (h_proto) {
+	case ETH_P_IP:
+		proto = load_byte(skb,
+				  ETH_HLEN + offsetof(struct iphdr, protocol));
```
Comment (blanquicet) [inline pkg/gadgets/trace/dns/tracer/bpf/dns.c:119]: nit: This could be taken from `iph->protocol`

**Review by blanquicet — APPROVED** [review]
LGTM!

File: pkg/gadgets/trace/dns/tracer/bpf/dns.c [hunk 2]
```diff
@@ -109,23 +107,85 @@ SEC("socket1")
...
-	__u16 id;
+	__u16 sport, dport, l4_off, dns_off, h_proto, id;
+	__u8 proto;
 	int i;
 
-	// Skip non-IP packets
-	if (load_half(skb, offsetof(struct ethhdr, h_proto)) != ETH_P_IP)
-		return 0;
+	// Do a first pass only to extract the port and drop the packet if it's not DNS
+	h_proto = load_half(skb, offsetof(struct ethhdr, h_proto));
```
Comment (alban) [inline pkg/gadgets/trace/dns/tracer/bpf/dns.c:115]: Note that I will remove the `load_half()` definition in #2579.

**Comment by mauriciovasquezbernal** [comment]
> LGTM! I was wondering if we could add some tests for this in the CI.

I fear it's too complicated. We'll need to have a cluster with ipv6 enabled, and/or being able to run containers with ipv6. I can open an issue to handle this later on.

**Pull request merged**